test/src/test/java/org/springframework/security/test/support/JwtAuthenticationBuilderTests.java

```diff
@@ -0,0 +1,133 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.Test;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtClaimNames;
+import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
+
+/**
+ * @author Jérôme Wacongne &lt;ch4mp@c4-soft.com&gt;
+ * @since 5.2
+ */
+public class JwtAuthenticationBuilderTests {
+	static class TestJwtAuthenticationBuilder extends JwtAuthenticationBuilder<TestJwtAuthenticationBuilder> {
+	}
+
+	@Test
+	public void defaultNameAndAuthority() {
+		final JwtAuthenticationToken actual = new TestJwtAuthenticationBuilder().build();
+
+		assertThat(actual.getName()).isEqualTo("user");
+		assertThat(actual.getAuthorities()).containsExactly(new SimpleGrantedAuthority("ROLE_USER"));
+	}
+
+	@Test
+	public void defaultNameAndRoleOverides() {
+		assertThat(new TestJwtAuthenticationBuilder().name("ch4mpy").build().getName()).isEqualTo("ch4mpy");
+		assertThat(new TestJwtAuthenticationBuilder().authority("TEST").build().getAuthorities())
+				.containsExactly(new SimpleGrantedAuthority("TEST"));
+		assertThat(new TestJwtAuthenticationBuilder().role("TEST").build().getAuthorities())
+				.containsExactly(new SimpleGrantedAuthority("ROLE_TEST"));
+	}
+
+	@Test
+	public void authenticationNameAndTokenSubjectClaimAreSet() {
+		final JwtAuthenticationToken actual = new TestJwtAuthenticationBuilder().name("ch4mpy").build();
+
+		assertThat(actual.getName()).isEqualTo("ch4mpy");
+		assertThat(actual.getTokenAttributes().get(JwtClaimNames.SUB)).isEqualTo("ch4mpy");
+	}
+
+	@Test
+	public void tokenIatIsSetFromClaims() {
+		final Jwt actual = new TestJwtAuthenticationBuilder().name("ch4mpy")
+				.claim(JwtClaimNames.IAT, Instant.parse("2019-03-21T13:52:25Z"))
+				.build()
+				.getToken();
+
+		assertThat(actual.getIssuedAt()).isEqualTo(Instant.parse("2019-03-21T13:52:25Z"));
+		assertThat(actual.getExpiresAt()).isNull();
+		assertThat(actual.getClaimAsInstant(JwtClaimNames.IAT)).isEqualTo(Instant.parse("2019-03-21T13:52:25Z"));
+		assertThat(actual.getClaimAsInstant(JwtClaimNames.EXP)).isNull();
+	}
+
+	@Test
+	public void tokenExpIsSetFromClaims() {
+		final Jwt actual = new TestJwtAuthenticationBuilder().name("ch4mpy")
+				.claim(JwtClaimNames.EXP, Instant.parse("2019-03-21T13:52:25Z"))
+				.build()
+				.getToken();
+
+		assertThat(actual.getIssuedAt()).isNull();
+		assertThat(actual.getExpiresAt()).isEqualTo(Instant.parse("2019-03-21T13:52:25Z"));
+		assertThat(actual.getClaimAsInstant(JwtClaimNames.IAT)).isNull();
+		assertThat(actual.getClaimAsInstant(JwtClaimNames.EXP)).isEqualTo(Instant.parse("2019-03-21T13:52:25Z"));
+	}
+
+	@Test
+	public void scopeClaimAreAddedToAuthorities() {
```

Comment:
The test name changed to `convertedClaimsAreMergedWithAuthorities` as what I test here is that converter is actually triggered and all authorities are merged.

> applications are going to work directly with GrantedAuthority instances, not with the scope claim

I think it is a dangerous assumption as applications might work with claims directly. Any claim. This can be achieved using SPeL or casting `Authentication`. I'm talking about use cases similar to what is done in `OAuth2ResourceServerController` from `spring-security-samples-boot-oauth2resourceserver-opaque` (I know JWT are not involved there, but I'm sure you can transfer the use-case)

Reply:
@jgrandja @rwinch -

The question in my mind is: What is the user's expectation when doing this:

```java
JwtAuthenticationToken token  = new JwtAuthenticationBuilder()
    .claim("scp", "read write").build();
token.getAuthorities().forEach(System.out::println)
```

What would the user expect to be displayed? Would it be zero, one, or two authorities?

Likewise, what would the user expect here:

```java
JwtAuthenticationToken token  = new JwtAuthenticationBuilder()
    .authorities(new SimpleGrantedAuthority("SCOPE_read")).build();
System.out.println(token.getClaims().get("scp"));
```

Note that there may be some value in changing out `roles(String...)` in this PR for `scopes(String...)` and that there is precedent there for adding those Strings as GrantedAuthorities, though I'd like your feedback also on the value of somehow adding those as a claim.

I have my own opinion on each of these (hopefully my earlier comments don't color your opinions too much), but I'd like to hear your feedback before proceeding.

Reply:
Plz forgive me answering where I'm not pinged, just providing the behavior I'd expect:
  * `.claim("scp", "read write")` behavior should depend on the registered authorities converter. So if default untouched (`JwtGrantedAuthoritesConverter`), authorities would be "SCOPE_read" and "SCOPE_write"
  * as converters are not bi-directional, `.authorities(new SimpleGrantedAuthority("SCOPE_read"))` should not modify claims, unless we add the ability to register a converter from authorities to claims, which seams overkill.

If a user wants claims to be set, he sets it and will get authorities updated if some of it are authorities related.

Reply:
You make a good point about the `opaque` use case, and that is something we are in the process of cleaning up in https://github.com/spring-projects/spring-security/issues/6498. Regardless, the canonical place to read authorities is from `Authentication`, not from something OAuth 2.0-specific.

I've had a quick chat with Rob offline about this to get his perspective (which I'm mentioning so it's clear why I'm responding without any posted comments from him or Joe).

As you have already alluded to, the transformation of the scope claim to a collection of GrantedAuthorities may change quite a bit from app to app. The simplest way for this API to support that is to keep them separate. The application can register a claim to the `Jwt` and an authority to the `JwtAuthenticationToken` to get the behavior they are wanting. 

Of course, if we need to provide support to sync these up down the road, then we can add it.

Reply:
>  the canonical place to read authorities is from Authentication

Agree

> The simplest way for this API to support that is to keep them separate

Disagree. The simplest way is to use the converter that is responsible for that in the app.

> The application can register a claim to the Jwt and an authority to the JwtAuthenticationToken

Do you really want to write the doc stating "You can add the claims you like. You also can build a test JwtAuthenticationToken by building a fully configured Jwt and providing it to the test `JwtAuthenticationToken`. Oh, by the way, in both cases you also need to hand configure the authorities afterward because we thought, in case you're not using default one, it was too complicated for you to provide the authorities converter to the test framework ?

Reply:
I think I can now put the answers to your questions

> `.claim("scp", "read write")` if default converter is not replaced, definitely **["SCOPE_read", "SCOPE_write"]**. Rule being all claims should be processed by authorities converter.

> `.authorities(new SimpleGrantedAuthority("SCOPE_read"))` should have no impact on claims. As converters are uni-directional, there is no reliable way to figure out claims from authorities using tested app converters. If a user wants claims to be set then he sets it (with an exception for subject claim)

> scope claim to a collection of GrantedAuthorities may change quite a bit from app to app

I would rather say that for Jwt, scope is a detail, if not a poison. In our discutions, it's something we should carefully avoid (you were right to ask me remove it from builder API).

As far as I (now) understand the spec and spring implementation, there is no requirement for an app to use neither "scope" nor "scp" claims to provide resource-server with authorities. It perfectly makes sens to use stuff like "authorities" or "roles" or whatever, choose a fancy serialization or decide they just don't want the "SCOPE_" prefix.

As a consequence, I think we **really have to keep the ability to configure the `JwtAuthenticationTokenBuilder` with the app authorities converter**
* defaulting the converter to `JwtGrantedAuthoritiesConverter`, most users won't even notice it's there and the rare ones replacing it in their app will have no difficulty to provide it to the test framework
* with very little work on the configuration lib, you can define this authorities converter as a been which could then be just `@Autowired`

Reply:
@jzheaux In response to your [comment](https://github.com/spring-projects/spring-security/pull/6748#discussion_r278233698):

> `JwtAuthenticationBuilder().authorities()`

I would expect this to add `GrantedAuthority`(s) to `JwtAuthentication`

> `JwtAuthenticationBuilder().claim()`

Here I would expect claim(s) to be added to `JwtAuthentication.Jwt.claims`. I **do not** see `GrantedAuthority`(s) being added in addition.

I know the default behaviour of `JwtAuthenticationConverter` doesn't align with my statement here but IMO these are 2 different use cases.

`JwtAuthenticationBuilder` simply builds a `JwtAuthentication`, pretty straight forward. However, `JwtAuthenticationProvider` which composes `JwtAuthenticationConverter` provides more complex behaviour as it's responsible for authenticating a `Jwt` into a `JwtAuthentication`.

Reply:
I think the documentation would read something like:

```
You can specify `GrantedAuthority`s the same way as other mock 
`Authentication` principals by calling the `authorities` method: {example} 

As a convenience, you could instead call `scopes`, which works the 
same as `roles`, just with the `SCOPE_` prefix instead of `ROLE_`: {example}
```

Since `GrantedAuthority` is the first-class citizen here, I doubt the docs would take the time to list the myriad ways the code chooses not to represent authorities.

Developers are smart - they can create their set of authorities in myriad ways and those that want to use a converter still can.

A very simple way that already exists today is `authentication(JwtAuthenticationToken)`. For example, they can build a `Jwt`, wire their converter into their test and construct the token themselves, with only a view lines of code:

```java
Jwt jwt = jwtConstruction();
Collection<GrantedAuthority> authorities = autowiredConverter.convert(jwt);
JwtAuthenticationToken token = new JwtAuthenticationToken(jwt, authorities);
this.mvc.perform(get().with(authentication(token)));
```

It's good to start with a minimal API and see what patterns emerge from the community at large. Really, there's nothing that says that something like `authoritiesConverter` can't be added later on.